### PR TITLE
initial Graphprot branch integration

### PR DESCRIPTION
**Pull request by tzk**
contains:

* graphprot model driver scripts
* graphprot input prep scripts
* additional tests
* improved trapezoidal reweighting

### CODE REVIEW

**Comment by bgruening** [comment]
Coooool, will review today!

File: eden/modifier/graph/vertex_attributes.py [hunk 1]
```diff
@@ -105,36 +105,183 @@ def discretize(graph_list=None, output_attribute='value', input_attribute='weigh
...
+    ...                            interpolate_up_end=3,
+    ...                            interpolate_down_start=5,
+    ...                            interpolate_up_start=1,
+    ...                            interpolate_down_end=7)
+    0
+    """
+    if pos <= interpolate_up_start:
+        """
+           ___
+        __/   \__
```
Comment (bgruening) [inline eden/modifier/graph/vertex_attributes.py]: ASCII art for the win!

File: graphprot/__init__.py [hunk 2]
```diff
@@ -0,0 +1,7 @@
+def thereisuinfun(string):
+    """A simple test function for testing unittests and module level tests.
```
Comment (bgruening) [inline graphprot/__init__.py]: is this still needed?

File: graphprot/graphprot_seqmodel [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+from sklearn.linear_model import SGDClassifier
+from eden.graph import Vectorizer
+from eden.model_base import ModelInitializerBase, main_script
+
+
+class ModelInitializer(ModelInitializerBase):
+
+    def _load_data(self, input_file):
+        """Load sequence tuples from file."""
+        from eden.converter.fasta import fasta_to_sequence
```
Comment (bgruening) [inline graphprot/graphprot_seqmodel]: I'm not a fan of imports at a function level. There are only a few cases where this is accepted and I think this load_data function is substantial for the program. So I would move it up to the top.

File: graphprot/graphprot_seqmodel [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        vectorizer_parameters = {
+            'r': [1, 0, 2, 3, 4],
+            'd': [4, 0, 1, 2, 3, 4, 5, 6],
+            'nbits': [14, 16, 18]}
+        return vectorizer, vectorizer_parameters
+
+    def estimator_init(self, n_iter):
+        """Setup the estimator and set of matching parameter choices."""
+        estimator = SGDClassifier(average=True, class_weight='auto', shuffle=True)
+        estimator_parameters = {'n_iter': randint(5, 200, size=n_iter),
```
Comment (bgruening) [inline graphprot/graphprot_seqmodel]: To make it easier for testing it would be nice to be able to put in a fixed seed with andom.seed().
Somewhere, I guess this can be done on a higher level not here in a function.

File: graphprot/prepare_graphprot_seqs.py [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+    help="Path to file containing chromosome limites as required by bedtools. Use this parameter disables automatic lookup via the genome id.")
+parser.add_argument(
+    "--negative_site_candidate_regions_fn",
+    help="Path to regions considered for placement of negatives in bed format")
+parser.add_argument(
+    "-v", "--verbose",
+    help="Be verbose.",
+    action="store_true")
+parser.add_argument(
+    "-d", "--debug",
```
Comment (bgruening) [inline graphprot/prepare_graphprot_seqs.py]: Where is the difference between -v and -debug? You could also use different debugging levels if you want with a single parameter ... for example. -v and -vv and -vvv

File: graphprot/prepare_graphprot_seqs.py [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+flank_length = args.seq_length - args.core_length
+flank_upstream_length = int(flank_length / 2)
+flank_downstream_length = int(flank_length / 2) + (flank_length % 2)
+if (args.core_length + flank_upstream_length + flank_downstream_length != args.seq_length):
+    raise Exception("Error: bad length calculation.")
+
+
+def dbg_head(sites, description="", n=npeek, run=args.debug):
+    """Print the first few bed entries."""
+    if (run):
```
Comment (bgruening) [inline graphprot/prepare_graphprot_seqs.py]: no brackets

File: graphprot/prepare_graphprot_seqs.py [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+    feature.name = prefix + feature.name
+    return feature
+
+
+def offset_zero_by_one(feature):
+    """Sets the start coordinate to 1 if it is actually 0.
+
+    Required for the flanking to work properly in those cases.
+    """
+    if (feature.start == 0):
```
Comment (bgruening) [inline graphprot/prepare_graphprot_seqs.py]: no brackets

File: graphprot/prepare_graphprot_seqs.py [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+        feature.start = feature.start + 1
```
Comment (bgruening) [inline graphprot/prepare_graphprot_seqs.py]: feature.start += 1

File: graphprot/prepare_graphprot_seqs.py [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+    return feature
+
+
+def get_flanks(cores,
+               flank_upstream_length=flank_upstream_length,
+               flank_downstream_length=flank_downstream_length):
+    """Calculate flanking regions of a core region."""
+    if (args.chromosome_limits is not None):
```
Comment (bgruening) [inline graphprot/prepare_graphprot_seqs.py]: brackets?

File: graphprot/prepare_graphprot_seqs.py [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+            r=0,
+            genome=args.genome_id).saveas()
+        # get downstream flanks
+        flanks_downstream = cores.flank(
+            s=True,
+            r=flank_downstream_length,
+            l=0,
+            genome=args.genome_id).saveas()
+    # check if sites and flanks have the same number of entries
+    if (not cores.count() == flanks_upstream.count() == flanks_downstream.count()):
```
Comment (bgruening) [inline graphprot/prepare_graphprot_seqs.py]: This looks wired. You want to have the not infront of the brackets? Is working, but I think this syntax is not really educative.

File: graphprot/prepare_graphprot_seqs.py [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+    fup_seq_fn = flanks_upstream.seqfn
+    cores_seq_fn = cores.seqfn
+    fdown_seq_fn = flanks_downstream.seqfn
+    viewpointfa = open(viewpointfa_fn, "wb")
+    with open(fup_seq_fn, "rb") as fup_tabseq, open(cores_seq_fn, "rb") as core_tabseq, open(fdown_seq_fn, "rb") as fdown_tabseq:
+        fup_reader = reader(fup_tabseq, delimiter="\t")
+        core_reader = reader(core_tabseq, delimiter="\t")
+        fdown_reader = reader(fdown_tabseq, delimiter="\t")
+        for fup, core, fdown in izip(fup_reader, core_reader, fdown_reader):
+            if (not fup[0] == core[0] == fdown[0]):
```
Comment (bgruening) [inline graphprot/prepare_graphprot_seqs.py]: see above

File: graphprot/graphprot_seqmodel (hunk 3 above)
Comment (tzk) [inline graphprot/graphprot_seqmodel]: True in general. I copied this from @fabriziocosta, currently all the model drivers do this. Plus point for doing it like this is that it simplifies copying functions between the different model drivers. We should change all or none of them for consistency.

File: graphprot/graphprot_seqmodel (hunk 4 above)
Comment (tzk) [inline graphprot/graphprot_seqmodel]: Definitively planned, will postpone this for when I actually start testing the script.

File: graphprot/graphprot_seqmodel (hunk 3 above)
Comment (fabriziocosta) [inline graphprot/graphprot_seqmodel]: I am for declaring everything at the beginning of the module.

File: graphprot/prepare_graphprot_seqs.py (hunk 5 above)
Comment (tzk) [inline graphprot/prepare_graphprot_seqs.py]: I'm using the EDeN logging helper now, this does not write errors to stderr but to stdout. @fabriziocosta, I think we should implement stderr/stdout discrimination in the eden.util configure_logging. Except you did this on purpose?

**Comment by tzk** [comment]
All done, thanks a lot for the review!

**Comment by bgruening** [comment]
Awesome! This looks great.
I have added https://github.com/fabriziocosta/EDeN/issues/58 to track the random-seed issue.

File: graphprot/prepare_graphprot_seqs.py (hunk 5 above)
Comment (fabriziocosta) [inline graphprot/prepare_graphprot_seqs.py]: yes, as in notebooks errors are rendered in pink, while I use the logger levels just to vary the amount of info displayed
